src/Firestore/Query.php

```diff
@@ -0,0 +1,477 @@
+     *
+     * Example:
+     * ```
+     * $query = $query->offset(10);
+     * ```
+     *
+     * @param int $number The number of results to skip.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function offset($number)
+    {
+        return $this->newQuery([
+            'offset' => $number
+        ]);
+    }
+
+    /**
+     * A starting point for the query results.
+     *
+     * Example:
+     * ```
+     * $query = $query->startAt($cursor);
+     * ```
+     *
+     * @param array $cursor A list of values.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function startAt(array $cursor)
+    {
+        return $this->newQuery([
+            'startAt' => [
+                'before' => true,
+                'values' => $this->valueMapper->encodeValues($cursor)
+            ]
+        ]);
+    }
+
+    /**
+     * A starting point for the query results.
+     *
+     * This method starts the result set AFTER the occurence of the given cursor.
+     *
+     * Example:
+     * ```
+     * $query = $query->startAfter($cursor);
+     * ```
+     *
+     * @param array $cursor A list of values.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function startAfter(array $cursor)
+    {
+        return $this->newQuery([
+            'startAt' => [
+                'before' => false,
+                'values' => $this->valueMapper->encodeValues($cursor)
+            ]
+        ]);
+    }
+
+    /**
+     * An end point for the query results.
+     *
+     * Example:
+     * ```
+     * $query = $query->endBefore($cursor);
+     * ```
+     *
+     * @param array $cursor A list of values.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function endBefore(array $cursor)
+    {
+        return $this->newQuery([
+            'endAt' => [
+                'before' => true,
+                'values' => $this->valueMapper->encodeValues($cursor)
+            ]
+        ]);
+    }
+
+    /**
+     * An end point for the query results.
+     *
+     * This method ends the result set AFTER the occurence of the given cursor.
+     *
+     * Example:
+     * ```
+     * $query = $query->endAt($cursor);
+     * ```
+     *
+     * @param array $cursor A list of values.
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function endAt(array $cursor)
+    {
+        return $this->newQuery([
+            'endAt' => [
+                'before' => false,
+                'values' => $this->valueMapper->encodeValues($cursor)
+            ]
+        ]);
+    }
+
+    /**
+     * Resets the Query.
+     *
+     * Example:
+     * ```
+     * $query = $query->clearQuery();
+     * ```
+     *
+     * @return Query A new instance of Query with the given changes applied.
+     */
+    public function clearQuery()
```

Comment:
Is there a benefit over just creating a new Query? I am little concerned about adding methods to just one language as it drives demand to add these to the other SDKs, which may not be what we want.